src/sync/index.js

```diff
@@ -0,0 +1,344 @@
+function sequentialLoop(iterations, process, exit){
+  var index = 0, done=false, shouldExit=false;
+
+  var loop = {
+      next:function(){
+        if(done){
+          if(shouldExit && exit){
+            return exit();
+          }
+        }
+
+        if(index < iterations){
+          index++;
+          process(loop);
+        }else{
+          done = true;
+
+          if(exit){
+            exit();
+          }
+        }
+      },
+
+      iteration:function(){
+        return index - 1; // Return the loop number we're on
+      },
+
+      break:function(end){
+        done = true;
+        shouldExit = end;
+      },
+  };
+  loop.next();
+  return loop;
+}
+
+function sync(db){
+  qclient.getBlockCount()
+    .then(
+      (value)=>{
+        currentBlockChainHeight = value - 1;
+        var options = {
+          "limit": 1,
+          "sort": [["blockNum", 'desc']]
+        }
+
+        db.Blocks.find({}, options).toArray(function(err, blocks){
+          var start_block = contract_deployed_block_num;
+          if(blocks.length > 0){
+            start_block = Math.max(blocks[0].blockNum + 1, start_block);
+          }
+
+          var initialSync = sequentialLoop(Math.ceil((currentBlockChainHeight-start_block)/batch_size), function(loop){
+            var end_block = Math.min(start_block + batch_size, currentBlockChainHeight);
+            var syncTopic = false, syncCOracle = false, syncDOracle = false, syncVote = false,
+                syncOracleResult = false, syncFinalResult = false;
+
+            // sync TopicCreated
+            contractEventFactory.searchLogs(start_block, end_block, Contracts.EventFactory.address, [Contracts.EventFactory.TopicCreated])
+              .then(
+                (result) => {
+                  console.log(`${start_block} - ${end_block}: Retrieved ${result.length} entries from TopicCreated`);
+                  // write to db
+                  _.forEach(result, (event, index) => {
+                    let blockNum = event.blockNumber;
+                    _.forEachRight(event.log, (rawLog) => {
+                      if(rawLog['_eventName'] === 'TopicCreated'){
+                        var topic = new Topic(blockNum, rawLog);
+                        db.Topics.insert(topic.translate());
+                      }
+                    })
+                  });
+
+                  syncTopic = true;
+
+                  if (syncTopic && syncCOracle && syncDOracle && syncVote && syncOracleResult && syncFinalResult){
+                    for(var i=start_block; i<=end_block; i++) {
+                      db.Blocks.insert({'blockNum': i});
+                    }
+                    start_block = end_block+1;
+                    loop.next();
+                  }
+                },
+              (err) => {
+                console.log(err.message);
+                syncTopic = true;
+                if (syncTopic && syncCOracle && syncDOracle && syncVote && syncOracleResult && syncFinalResult){
```

Comment:
is this necessary because all the searchlogs are asynchronous?

Reply:
+1

Reply:
yes. Note that err neednt due to code error. A parsing error caused by legacy events cant be ignored. So I still mark this as true and keep looping